src/scope.rs

```diff
@@ -45,6 +46,9 @@ pub struct Scope<'a> {
     files: Set<&'a Path>,
     /// All defined or directly imported functions
     directly_imported_or_defined: Set<UniCase<&'a str>>,
+    /// Directly imported definitions grouped by file. This introduces denormalized
+    /// data between directly_imported and this field.
+    definitions_per_import: Map<&'a syntax::Import, Set<UniCase<&'a str>>>,
```

Comment:
This struct is growing a little bit too... organically.

What do you think about instead of just having sets, keep a `Map` from name to "Item", where "Item" would contain info about the origin? (It could be a tuple, struct, or perhaps just `&Path`).

I think, this way we could get rid of all these `directly`/indirectly monstrocities (and just check whether the "origin" exists inside the direct-import list of the current file)

Reply:
I was thinking about file graph here tbh

Reply:
And then in scope.rs, instead of `for path in files.keys() {` we would DFS or whatever.

Reply:
Yup, a graph it could be (it's kind of a graph now, but anyway). But still, we need to have set of items in scope, preferably marked with their directness and origin and I think a simple `Map<Unicase<&str>, &Path>` would be a lot better than current "just make a 5 sets and a map and make sure you don't mess up your `extends`".